### ISSUE TITLE
The future of this library: Improve bootstrapping, initialization, and remove common developer errors

### ISSUE THREAD

**Issue by bendiknesbo** [coopnorge/go-datadog-lib#418]
In a meeting between me, @nhhagen and @hawi74, we discussed some problems with the current state of the library.

- The library requires verbose initialization (many lines of code to start the datadog-collector)
- The library is prone to errors: Middlewares needs to be initialized **after** the datadog-collector is started
- The Tracer&Profiler is started separately from the Metrics-collector
- The config is required to be read and parsed by the consuming library (from env-vars or env-files) and then passed to the library
- Some functionality might still require direct calls to the underlying library (such as starting/ending manual spans)
- Some parts have added null-safe checks (e.g. creating gRPC middleware when datadog is disabled), but others haven't (such as the metrics-collector), causing runtime-panics if the config is invalid.
- When an application shuts down, datadog should attempt to flush any remaining traces/profiling/metrics before shutting down.

We discussed creating a v3 of the go-datadog-lib (or a new lib like go-monitoring-lib or go-telemetry-lib), or if we should only add a new way to bootstrap everything for datadog and deprecate the old methods.

We discussed 3 main ways of using the library:
1. As a config-only-lib (expect users to use dd-trace-go to manually start new traces/spans, and use contrib-middleware where needed)
2. As a fully encapsulated monitoring library (hide all implementations, wrap everything) (go-monitoring-lib)
3. Move to open-telemetry SDK instead of dd-trace-go, combined with one of the two above approaches (go-telemetry-lib)

**Comment by bendiknesbo** [comment]
The consensus from the meeting was that we should continue using dd-trace-go, but encapsulate more of the "read config and start collectors"-part, and ensure a better developer-experience for consumers.

- We want to provide a single entrypoint to the library that reads the config from environment-variables.
- We want to provide value-adding wrappers for existing middlewares where possible (good default-values, no-op middlewares if datadog is disabled, ensuring the datadog collectors is started)
- We want to provide a better functionality when shutting down the application, to try to flush remaining traces/profiling/metrics.

We also want some small features/bug-fixes:
- We want to automatically add the tracelogger hook to the global go-logger instance (for trace->log correlation)
- We want to hijack the following log-message `Datadog Tracer v1.68.0 WARN: Unsupported protocol "" in Agent URL "/var/run/datadog/apm.socket". Must be one of: http, https, unix.` and swallow that message, as it is causing a lot of confusion for developers.

**Comment by hadrienk** [comment]
I think we should evaluate the **option 3**. Many problems that are highlighted would go away: 

> We want to provide a single entrypoint to the library that reads the config from environment-variables.

The entire open telemetry ecosystem is configurable using env: https://opentelemetry.io/docs/specs/otel/configuration/sdk-environment-variables/. Effectively making it a zero-setup API, if one choses to do so.

>  no-op middlewares if datadog is disabled

Very important indeed, this is covered: https://opentelemetry.io/docs/specs/otel/metrics/noop/

> ensuring the datadog collectors is started 

No need if using OTLP. The OTLP protocol also allows to buffer/aggregate/filter the metric/traces/logs using a very flexible architecture:
https://opentelemetry.io/docs/collector/

> trace->log correlation

Out of the box, especially now that logging is part of OTLP.

**Comment by bendiknesbo** [comment]
Good thoughts. I have treading a bit careful around open-telemetry in go, as it has been mentioned multiple times that they are known to make breaking changes.

See this Reddit-post from 2 years ago: https://www.reddit.com/r/golang/comments/yq3h5w/is_it_worth_instrumenting_with_opentelemetry/

And this from 1 year ago: https://www.reddit.com/r/golang/comments/18dn0ng/behavioral_breaking_change_proposal_in_otel_go/

It might be that they have matured more now.

**Comment by nhhagen** [comment]
@hadrienk and @tuhtah we also do not know how well the open-telemetry SDK works with Datadog. Replacing Datadog (the service) is not on the table as that is selected as the operational tool from the org.

For logging we still have to adhere to https://inventory.internal.coop/docs/default/component/guidelines/data_protection/#personal-data-in-logs

> Personal Data in logs[¶](https://inventory.internal.coop/docs/default/component/guidelines/data_protection/#personal-data-in-logs)
Personal Data MUST NOT be written to system logs. System logs are not considered to be a part of the system since the logs are sent to a central separate system to enable the required observability for operational use. For more information see, [Obeservability / Logging / Personal Data](https://inventory.internal.coop/docs/default/component/guidelines/observability/#personal-data).

**Comment by hadrienk** [comment]
> Good thoughts. I have treading a bit careful around open-telemetry in go, as it has been mentioned multiple times that they are known to make breaking changes.
> 
> See this Reddit-post from 2 years ago: https://www.reddit.com/r/golang/comments/yq3h5w/is_it_worth_instrumenting_with_opentelemetry/
> 
> And this from 1 year ago: https://www.reddit.com/r/golang/comments/18dn0ng/behavioral_breaking_change_proposal_in_otel_go/
> 
> It might be that they have matured more now.

Yes, I think that they took a 180 turn with some of the internals of the sdk at some point. But I wonder if this was also due to changes in the spec itself. The changes I've experienced previously were trivial I must say.  

https://github.com/open-telemetry/opentelemetry-go/blob/main/VERSIONING.md

**Comment by hadrienk** [comment]
> [@hadrienk](https://github.com/hadrienk) and [@tuhtah](https://github.com/tuhtah) we also do not know how well the open-telemetry SDK works with Datadog. 

True, we would have to try it to be sure, but they are [advertising it](https://docs.datadoghq.com/opentelemetry/interoperability/otlp_ingest_in_the_agent/?tab=host). The earliest mention is from one [year ago](https://www.datadoghq.com/blog/agent-otlp-log-ingestion/).
One area that might be tricky is the mappings, but it seems [well documented](https://docs.datadoghq.com/opentelemetry/schema_semantics/). 

> Replacing Datadog (the service) is not on the table as that is selected as the operational tool from the org.

Of course, using open telemetry does not mean replacing datadog. 

> For logging we still have to adhere to https://inventory.internal.coop/docs/default/component/guidelines/data_protection/#personal-data-in-logs
> 
> > Personal Data in logs[¶](https://inventory.internal.coop/docs/default/component/guidelines/data_protection/#personal-data-in-logs)
> > Personal Data MUST NOT be written to system logs. System logs are not considered to be a part of the system since the logs are sent to a central separate system to enable the required observability for operational use. For more information see, [Obeservability / Logging / Personal Data](https://inventory.internal.coop/docs/default/component/guidelines/observability/#personal-data).

Not sure what you mean here, the logging part I'm mentioning is merely a logger API details (log, slog, logrus, zerolog etc). It would send the logs to the same place they are sent today (or somewhere else if we decide so).

**Comment by bendiknesbo** [comment]
> Not sure what you mean here, the logging part I'm mentioning is merely a logger API details (log, slog, logrus, zerolog etc). It would send the logs to the same place they are sent today (or somewhere else if we decide so).

I haven't had a look at how OTEL does logging yet, but I assume it still supports logging via reading stdout/stderr, similar to how we do it today?

**Comment by hadrienk** [comment]
> > Not sure what you mean here, the logging part I'm mentioning is merely a logger API details (log, slog, logrus, zerolog etc). It would send the logs to the same place they are sent today (or somewhere else if we decide so).
> 
> I haven't had a look at how OTEL does logging yet, but I assume it still supports logging via reading stdout/stderr, similar to how we do it today?

It's pretty flexible, you can either export the logs using otlp, in that case there's no output. Or use a bridge to other structured log tech. I've tested with the new golang slog in a couple of other projects.

**Comment by nhhagen** [comment]
OpenTelemetry discussion moved to:
- https://github.com/coopnorge/engineering-issues/issues/431

**Comment by nhhagen** [comment]
@coopnorge/engineering closing this as resolved now :)
